### PR TITLE
chore(types): clean up TypeScript support

### LINKED ISSUE #3430: Better typings Support for the webdriver api
Feature Request
---

Better typings Support for the webdriver api

Currently many of the webdriver functions specified in the selenium api are not possible to use with the new protractor typings. I see why you might want to hide some of the api, however a lot is missing right now.

We upgraded to 4.0.2 and want to use the new typings however we are unable to as long as parts of the api is missing. (at least in a easy and clean way)

We specifically use 
```typescript
  executeScript: (script: string) => webdriver.promise.Promise<any[]>;
  manage: () => Options;
```

But more can be found her: http://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/index_exports_WebDriver.html

As a small note shouldn't 

```typescript
export class Webdriver {

}
```

be an interface instead? I believe an interface would be possible to extend without changes code in the protractor module.

Let me know if you are interested in a pull request

### LINKED ISSUE #3477: Property 'manage' doesn't exist on type 'ProtractorBrowser'
I am trying to use `browser.manage().window().maximize()` in my config and specs, The typescript compiler throws the above error. I think the method is not yet exported! Similar case is for `browser.switchTo().alert()` few of the properties are still not exported!

- Protractor Version: `4.0.3`

### LINKED ISSUE #3500: optional class members in protractor using typescript
Optional class members were introduced in `TypeScript 2.0` as a result older typescript `versions(1.8)` using new `Protractor 4.0.4` version would throw an compilation error : 
```
node_modules/protractor/built/index.d.ts(516,13): error TS1112: A class member cannot be declared optional.
```
due to this declaration in `index.d.ts` : 

screen shot 2016-08-27 at 9 32 33 pm

Suggestions:

* Either officially support protractor with typescript 2.0  'or'
* Switch to Interfaces having optional members rather than classes which would promote backward compatibility.

Any thoughts?

### PR DESCRIPTION
**Pull request by cnishina**
- fix exampleTypescript to have noGlobals
- change the exampleTypescript tsconfig.json to include @types
- alias promise as wdpromise everywhere
- refactor gulpfile task 'types' to use multiple files from tsc
- add browser method types applied from webdriver

this closes #3430, closes #3477, and closes #3500

### CODE REVIEW

**Comment by ZurabWeb** [comment]
+1

File: lib/browser.ts [hunk 1]
```diff
@@ -32,20 +33,36 @@ for (let foo in webdriver) {
   exports[foo] = webdriver[foo];
 }
 
-// Explicitly define webdriver.WebDriver.
+// Explicitly define webdriver.WebDriver
```
Comment (juliemr) [inline lib/browser.ts]: Please add a TODO to extend Webdriver from our typings for it.

File: lib/element.ts [hunk 2]
```diff
@@ -20,25 +20,25 @@ let WEB_ELEMENT_FUNCTIONS = [
 // Explicitly define webdriver.WebElement.
```
Comment (juliemr) [inline lib/element.ts]: Please add a TODO here as well

File: gulpfile.js [hunk 3]
```diff
@@ -97,90 +97,28 @@ gulp.task('tsc:globals', function(done) {
...
 });
 
 gulp.task('default',['prepublish']);
 
 gulp.task('types', function(done) {
-  var folder = 'built';
-  var files = ['browser', 'element', 'locators', 'expectedConditions',
-    'config', 'plugins', 'ptor'];
-  var outputFile = path.resolve(folder, 'index.d.ts');
+  var outputFile = path.resolve('built', 'index.d.ts');
```
Comment (juliemr) [inline gulpfile.js]: Instead of having a gulp task for this, can this now just be a `lib/index.ts` file?

Comment (juliemr) [inline gulpfile.js]: Or do we still need this to handle the reference path business?

File: lib/globals.d.ts [hunk 4]
```diff
@@ -25,7 +25,7 @@ declare namespace NodeJS {
     Command: any;
```
Comment (juliemr) [inline lib/globals.d.ts]: Can you add a file comment to the top of this file explaining what it's for? Especially with `globals.ts` floating around, it's confusing.

**Comment by juliemr** [comment]
Sorry this took me so long to review. Shame!

I think we could use some more documentation to make navigating our TS infrastructure easier, so I requested a bit here. Feel free to leave everything else for follow-up PRs.

File: lib/globals.d.ts (hunk 4 above)
Comment (cnishina) [inline lib/globals.d.ts]: Done. Also cleaning up the rest of the webdriver types

File: gulpfile.js (hunk 3 above)
Comment (cnishina) [inline gulpfile.js]: Created a `lib/index.d.ts` instead of generating it.

File: lib/element.ts (hunk 2 above)
Comment (cnishina) [inline lib/element.ts]: done.

File: lib/browser.ts (hunk 1 above)
Comment (cnishina) [inline lib/browser.ts]: done.

**Comment by juliemr** [comment]
Looks good.

File: lib/globals.d.ts (hunk 4 above)
Comment (juliemr) [inline lib/globals.d.ts]: I was thinking something at the top of the file to describe why the entire file exists (e.g. ambient typings that aren't captured in the rest of our type dependencies, but we can revisit this.

**Pull request merged**